packages/tldraw/src/lib/shapes/shared/useAsset.ts

```diff
@@ -1,38 +1,104 @@
-import { TLAssetId, TLShapeId, useEditor, useValue } from '@tldraw/editor'
-import { useEffect, useState } from 'react'
+import {
+	MediaHelpers,
+	TLAssetId,
+	TLImageShape,
+	TLShapeId,
+	TLVideoShape,
+	useEditor,
+	useReactor,
+} from '@tldraw/editor'
+import { useRef, useState } from 'react'
 
-/** @internal */
-export function useAsset(shapeId: TLShapeId, assetId: TLAssetId | null, width: number) {
+function useScalableAsset(shapeId: TLShapeId, assetId: TLAssetId | null) {
 	const editor = useEditor()
-	const [url, setUrl] = useState<string | null>(null)
 	const asset = assetId ? editor.getAsset(assetId) : null
-	const culledShapes = editor.getCulledShapes()
-	const isCulled = culledShapes.has(shapeId)
-
-	const shapeScale = asset && 'w' in asset.props ? width / asset.props.w : 1
-	// We debounce the zoom level to reduce the number of times we fetch a new image and,
-	// more importantly, to not cause zooming in and out to feel janky.
-	const screenScale = useValue('zoom level', () => editor.getZoomLevel() * shapeScale, [
-		editor,
-		shapeScale,
-	])
-
-	useEffect(() => {
-		if (isCulled) return
-
-		let isCancelled = false
-		const timer = editor.timers.setTimeout(async () => {
-			const resolvedUrl = await editor.resolveAssetUrl(assetId, {
-				screenScale,
-			})
-			if (!isCancelled) setUrl(resolvedUrl)
-		}, 500)
-
-		return () => {
-			clearTimeout(timer)
-			isCancelled = true
-		}
-	}, [assetId, asset?.props.src, isCulled, screenScale, editor])
-
-	return { asset, url }
+
+	// The resolved url of the asset that should be used by the shape (duplicated as a ref)
+	const [state, setState] = useState<{ url: string | null; isAnimated: boolean }>({
+		url: null,
+		isAnimated: false,
+	})
+
+	const rPrevUrl = useRef(state.url)
+	// The previous screen scale at which we requested a new url
+	const rPrevScreenScale = useRef(0)
+	// A timeout we can clear if we need to request a new url
+	const rTimeout = useRef<any>(-1)
+
+	useReactor(
+		'update url',
+		() => {
+			const shape = editor.getShape<TLImageShape | TLVideoShape>(shapeId)
+			if (!shape) return
+
+			const asset = assetId ? editor.getAsset(assetId) : null
+			if (!asset) return
+
+			if (editor.getCulledShapes().has(shapeId)) return
+
+			const screenScale =
+				editor.getZoomLevel() * ('w' in asset.props ? shape.props.w / asset.props.w : 1)
```

Comment:
ugh, yeah. now i remember what was tickling my brain on why we needed to make this wait a tick on first page load :-/ (and why we were thinking of some kind of `onBeforeMount`).

the problem is that the board, when it loads, sets the zoom twice, in a multiplayer room. it first loads as a zoom of 1, but then, after reading the url it sets the desired zoom. so it would make _all_ the images in the room load twice at the moment if we didn't make this wait a tick. UGH.

thoughts?